### ISSUE TITLE
make it easier to instantiate a model

### ISSUE THREAD

**Issue by NickleDave** [vocalpy/vak#362]
A key need for users is to get a quick sense of how a model is performing.  
To do this, they need to load a model and inspect its outputs.

Due to the current design of the API, it's both complicated and unclear how to make an instance of a model, and then load a checkpoint into that model.

Intuitively, you'd think you would be able to just say, e.g.
```python
model = TweetyNetModel()
model.load(checkpoint_path)
```

However this fails with:
```python
TypeError: __init__() missing 4 required positional arguments: 'network', 'loss', 'optimizer', and 'metrics'
```

The "correct" way is to use the `from_config` class method, but even knowing this, it's still annoying to use.
That's because this method does not just accept, e.g., a path to a config file, but instead requires a dictionary `config` that maps a set of keys, `{'network', 'optimizer', 'loss'}` to dictionaries that are themselves key-value pairs mapping hyperparameter names to their values.  

How do I get such a dictionary, you might ask?

Even more counterintuitively, you cannot just pass the whole config.toml file to `vak.parse.config` and get back a model config `dict`. Although functions for parsing model-specific tables in a .toml config file live within the `config` module, they are not called directly by `vak.parse.config`. Instead you have to call those functions directly, passing in both the path to the config.toml file and a list of mdoel names.  
Accordingly, all the `cli` functions ('eval', 'predict', 'train', etc.) do something like this to get the model config:
```python
model_config_map = vak.config.models.map_from_path(config_gr41rd51, config.learncurve.models)
models_map = vak.models.from_model_config_map(
    model_config_map,
    num_classes=len(labelmap),
    input_shape=val_dataset.shape[1:],
)
```
This is both verbose and non-obvious.

A relatively quick fix would be to add another class method, `from_toml_path`, that does all of this work for a user.

It's still annoying that you can't just say `model = ModelClass()' though.

Fixing that would require a breaking change, to something like a pytorch-lightning model, where the person designing the model specifies default hyperparameters, e.g. through methods like `get_default_loss`, but those can optionally be changed by passing in configurations as kwargs dicts. For my taste it would be possibly cleaner if defaults are just specified as class-level variables instead of requiring users to write methods which would then obscure the default hyperparameters--i.e. with pytorch-lightning you can't easily "introspect" the model to find out that the default learning rate is 0.001 because it's a magic number assigned within a method.

**Comment by NickleDave** [comment]
so a new, breaking-change `Model` would look like:
```
import torch.nn
import torch.optim

import vak.Model

class FakeNetModel(vak.Model):
    Network = FakeNetwork
    default_network_config = {}
    Optimizer = torch.optim.SGD
    default_optimizer_config = {}
    Loss = torch.nn.CrossEntropyLoss
    
    def __init__(self,
                 net_config=None,
                 opt_config=None,
                 loss_config=None):
        net_kwargs = net_config if net_config else self.default_network_config
        self.network = self.Network(**net_kwargs)
        ...
```

not sure if config validating / unpacking should live in base `Model` class, probably yes.  
Would need to do some sort of validation of any sub-class too. Tried to do this before, and I forget why I gave up :/

**Comment by NickleDave** [comment]
Will fix this by avoiding sub-classing altogether and just representing a `Model` as a dataclass basically, that gets passed into `Engine`, as described in #536 

This doesn't give us a `__call__` in the same way one can do with a `torch.Module` though :(

**Comment by NickleDave** [comment]
Ok, so a step towards fixing this :crossed_fingers: hopefully is using `lightning` as a backend, as done in #598.

What do we need to now:
- definitely remove `from_config` class method
- find a way to make it so when I instantiate `VakModel()` with no args, I get a thing that's really a `LightningModule` that has whatever network function I want associated with it
  - some sort of decorator?
  - my basic idea is to represent a *task* as a `LightningModule`, and let the network applied to that task vary, basically as described here: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#inference-in-production
    - the task can have a default `loss` and `optimizer`, removing the absolute requirement to pass those in and configure them 
- can we at the same time maintain the ability to pass in a `network_config`, `loss_config`, `optimizer_config` or is that too much?

**Comment by NickleDave** [comment]
This is in progress.

Putting thoughts here about how to achieve the main goal in this issue.  
I have pretty much implemented this but want to document the design decisions.  

- What are the bare minimum requirements to instantiate a model? Assume that we represent "windowed frame classification" as a task / family of models as in the comment above.
  - First of all, at the *model* level, we need a `labelmap`, a mapping from integer class outputs of the network to string labels used for annotation. The model as a whole can't do its job without this labelmap, i.e., it can't eval (need to compute string metrics) and it can't predict (convert network outputs to strings). So this has to be a required argument. 
    - Aside, this is an argument for developing the labelmap class as in #582 with a `from_labels` method so one can just do `labelmap = vak.LabelMap.from_labels('abcde', background=True)` (where `background=True` would be the default) instead of writing a `dict` by hand. This `labelmap` can have bonus methods like `num_classes` that one could use in an `__init__` as below.
  - For the network that "belongs" to a model that does windowed frame classification, there are two required arguments:
    - (1) the number of classes
        - Alternatively, we could have a required argument of the **network** (not just the model) be a labelmap that maps from the integer outputs of the model to the string labels (which may be whole numbers that are strings). This would allow directly mapping output of the network to strings. But this increases complexity and conflates two different levels of abstraction: the network is a math function, that maps each frame of continuous values to an integer class. The model does the work of converting this back to string annotations
    - (2) the input shape
        - For the network alone (again, not the model) we don't really need the window size, at least for TweetyNet
        - but it seems better to go with the agnostic `input_shape` for now since that argument can be consistent across models (although maybe having diff't semantic meaning for diff't networks is a bad idea)

So one can make a model in two steps--(1) instantiate network, (2) instantiate model, passing in the instance of the network--which is still easier to work with interactively than requiring a config. But I think we can make it even easier to instantiate a model by adding the required  arguments for the networks to the `__init__` of the superclass WindowedFrameClassificationModel.
Basically we do the following:
- ask for `labelmap` + `input_shape`
- require that a user either pass in a network instance, or a `labelmap` and an `input_shape`
    - If the user passes in the `labelmap` + `input_shape`, we instantiate the network with just those arguments

Like so:

```python
labelmap = vak.LabelMap.from_labels('abcde')
# WindowedFrameClassification.__init__ instantiates network for us
model = vak.models.TweetyNetModel(labelmap=labelmap, input_shape=(1, 513, 88))
model.load_state_dict_from_path(ckpt_path)
```

```python
labelmap = vak.LabelMap.from_labels('abcde')
# we want to instantiate by hand for some reason
net = vak.nets.TweetyNet(num_classes=len(labelmap), input_shape=(1, 513, 88), conv1_filters=16)
model = vak.models.TweetyNetModel(labelmap=labelmap, network=network)
model.load_state_dict_from_path(ckpt_path)
```

A separate question is if we could avoid needing the `labelmap` argument if we assume that it's part of a `post_tfm`. This could be true if the `post_tfm` returns both the vector of labeled frames / time bins with a clean-up step applied as well as the strings produced from the vector. But then there would be a lot of non-transparent magic happening inside the `eval` and `predict` functions. My gut feeling now is to keep the `labelmap` argument,  that gets turned into a transform `lbl_tb2labels` inside the `WindowedFrameClassification.__init__` and used in eval, and then allow for the optional `post_tfm` that is applied **only** to the network outputs, i.e. the labeled timebins/frame vector.

**Comment by NickleDave** [comment]
Think I closed this by accident :thinking: 

Further detail:  
can we share code between `eval` and `predict` using a `forward` method on the **model**?
This might be more trouble than it's worth

Note that `predict_step` of a `LightningModule` calls a `forward` on the module if it is implemented. (Not sure about `validation_step`). And in fact we could call it ourselves inside those classes when we over-ride them.

But what we also want to do is be able to say `out = Model(x)` and get full predictions. This is kind of abusing the API to make it do something else; we get a "free" forward method we can use on the model since during our `train_step` we call `forward` on the model's network directly. Ultimately this is what we need to be able to do to get a model where we can just pass in a tensor and get out predictions when working interactively

**Comment by NickleDave** [comment]
Closed by #605
